### PR TITLE
Add an aggregator for IPv4 and IPv6 subnets

### PR DESCRIPTION
**Pull request by salvatore-campagna**
This PR addresses a request for an IP address subnet aggregator (see #57964).

Parameters accepted by the aggregator include:
* `prefix_length` (integer, required): defines the _network_ size of the subnet mask
* `is_ipv6` (boolean, optional, default: false): defines whether the prefix applies to IPv6 (true) or IPv4 (false) IP addresses
* `min_doc_count` (integer, optional, default: 1): defines the minimum number of documents for a bucket to be returned in the results
* `append_prefix_length` (boolean, optional, default: false): defines if the prefix length is appended to the IP address key when returning results
* `keyed` (boolean, optional, default: false): defines whether the result is returned keyed or as an array of buckets

Each bucket returned by the aggregator represents a different subnet. IPv4 subnets also include a `netmask` field set to the subnet mask value (i.e. "255.255.0.0" for a /16 subnet).

Related to: https://github.com/elastic/elasticsearch/issues/57964 and https://github.com/elastic/kibana/issues/68424

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/TransformAggregations.java
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/TransformAggregations.java:68]: 👍

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by salvatore-campagna** [comment]
I am aware I need to add documentation about this new aggregation...I am just waiting for some feedback about the interface.

**Comment by flash1293** [comment]
The API makes a lot of sense to me - this would be a very good fit for how Kibana does things today 👍

**Comment by salvatore-campagna** [comment]
This is a sample request including an aggregation on an IPv4 field and an aggregation on an IPv6 field:
`curl -X GET "localhost:9200/network/_search?pretty" -H 'Content-Type: application/json' -d'
{
  "size": 0,
  "aggs": {
    "/16": {
      "ip_prefix": {
        "field": "ipv4",
        "prefix_len": 16,
        "is_ipv6": false,
        "min_doc_count": 1,
        "append_prefix_len": false,
        "keyed": false
      }
    },
    "/64": {
      "ip_prefix": {
        "field": "ipv6",
        "prefix_len": 64,
        "is_ipv6": true,
        "min_doc_count": 1,
        "append_prefix_len": false,
        "keyed": false
      }
    }
  }
}'`

And this is the response:

`{
  "took" : 29,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 5,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "/16" : {
      "buckets" : [
        {
          "key" : "192.168.0.0",
          "netmask" : "255.255.0.0",
          "doc_count" : 4,
          "is_ipv6" : false,
          "prefix_len" : 16
        },
        {
          "key" : "192.169.0.0",
          "netmask" : "255.255.0.0",
          "doc_count" : 1,
          "is_ipv6" : false,
          "prefix_len" : 16
        }
      ]
    },
    "/64" : {
      "buckets" : [
        {
          "key" : "2001:db8:85a3::",
          "doc_count" : 1,
          "is_ipv6" : true,
          "prefix_len" : 64
        },
        {
          "key" : "2001:db8:85ab::",
          "doc_count" : 1,
          "is_ipv6" : true,
          "prefix_len" : 64
        },
        {
          "key" : "2001:dc4:11b2:7::",
          "doc_count" : 1,
          "is_ipv6" : true,
          "prefix_len" : 64
        },
        {
          "key" : "2001:dc4:11b3::",
          "doc_count" : 1,
          "is_ipv6" : true,
          "prefix_len" : 64
        },
        {
          "key" : "2004:db3:85f4:10::",
          "doc_count" : 1,
          "is_ipv6" : true,
          "prefix_len" : 64
        }
      ]
    }
  }
}`

**Comment by salvatore-campagna** [comment]
NOTE: I put the code under `org.elasticsearch.search.aggregations.bucket.range` which might not be ideal since this is not really a range aggregation (like other range aggregations). Initially this was done supporting multiple ranges and I just left the code there. Looking at other packages I don't see a clear candidate for this...

**Comment by not-napoleon** [comment]
You can create a new package under `o.e.s.aggregations.bucket` for this agg.  That's a pretty standard pattern for us.

One thing we should probably test - what happens if we get mixed IPv4 and IPv6 addresses in the same run?  That doesn't seem like something the user would do on purpose, but for example if they're using an index pattern, it's not hard to imagine that the same field might have changed from IPv4 in one index to IPv6 in another.  We should do something sensible in that case, even if it's just failing the shard if the stored data doesn't match what the aggregation expected.

**Comment by salvatore-campagna** [comment]
I am adding a test were I index IPv6 addresses and then try to aggregate with `is_ipv6 = false` and viceversa, index documents with IPv4 addresses and try to aggregate with `is_ipv6 = true`. I think here the result is `unpredictable`. The issue is that all IP addresses are encoded as 16-bytes arrays. Then depending on whether the IP address is v6 or v4 the mask is applied "shifting" the actual mask by 12 bytes if the address is IPv4. I have been thinking of a clever way to handle this so that we don't need to know if the address is IPv4 or IPv6 but could not realise something good.

**Comment by salvatore-campagna** [comment]
I was wondering if we shoud, in some way, encode the IP version in the `IpFIeldMapper` (without having to distinguish two different types IPv4 and IPv6 but still keeping just IP as the data type). I am not sure this is doable.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (salvatore-campagna) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:23]: I will remove some od the documents here just to make the whole example more compact.

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:6]: I think the first sentence should have a stronger description of the aggregation. Something like:

`A bucket aggregation that groups documents based on the network or sub-network of an IP address.`

The second sentence has a subject-verb agreement error:
```suggestion
IP addresses can be aggregated according to their network or sub-network. An IP address consists of two groups of bits: the most significant bits which represent the network prefix, and the least significant bits which represent the host.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:9]: A heading here may make this more readable:
```suggestion
[[ipprefix-agg-ex]]
==== Example

For example, consider the following index:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:44]: Try to avoid future tense (_will_) when possible. I also split the first sentence so it's a little easier to read.
```suggestion
The following aggregation groups documents into buckets. Each bucket identifies a different sub-network. The sub-network is calculated applying a netmask with prefix length of `24` to each IP address in the `ipv4` field:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:156]: I don't know if we need this in a separate note. I'd consider moving it above the response.

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:157]: These examples are great. However, I think we should also define the input parameters and response properties in a list. For example, we don't state whether `field` needs to be an `ip` field or not.

You can use the adjacency matrix docs for an example:

https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-adjacency-matrix-aggregation.html#adjacency-matrix-agg-params

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:5]: Just adding some whitespace for readability.
```suggestion
++++

```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:160]: Some minor grammar fixes:
```suggestion
Set the `keyed` flag of `true` to associate an unique IP address key with each bucket and return sub-networks as a hash rather than an array:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:158]: If possible, try to set explicit anchor IDs for each heading. Otherwise, our docs build will auto-generate it, which can break the anchors and create namespacing issues.
```suggestion
[[ipprefix-agg-keyed-response]]
==== Keyed Response
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:219]:
```suggestion
[[ipprefix-agg-append-prefix-length]]
==== Append the prefix length to the IP address key

Set the `append_prefix_len` flag to `true` to catenate IP address keys with the prefix length of the sub-network:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:281]:
```suggestion
[[ipprefix-agg-min-doc-count]]
==== Minimum document count

Use the `min_doc_count` parameter to only return buckets with a minimum number of documents.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:101]: I'd remove the `see section below` text. This could easily break if we later remove or reorganize these docs.
```suggestion
To aggregate IPv6 addresses, set `is_ipv6` to `true`.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:334]: This would be great to note in the parameter and response property definitions.

**Review by jrodewig — COMMENTED** [review]
Thanks @salvatore-campagna! Great job!

I took a first pass and left some comments. The only large piece of feedback I had was that it'd be great to have formal parameter and response body definitions (similar to those for the [adjacency matrix agg](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-aggregations-bucket-adjacency-matrix-aggregation.html#adjacency-matrix-agg-params)).

I'd like to take another look once you address the comments and others have had a chance to review.

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (salvatore-campagna) [reply docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:6]: Yes, you are right...I thought about that. I think favouring conciseness assuming uses do know what we are talking about is important.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
I have just realised that having `min_doc_count` default to 0 might not be correct. Probably the default value should be 1. Actually the minimum allowed value for `min_doc_count` should be 1 instead of 0. Having `min_doc_count = 0` might suggest that we return empty buckets too in the response, which is not the case here.

**Comment by nik9000** [comment]
> I have just realised that having `min_doc_count` default to 0 might not be correct. Probably the default value should be 1. Actually the minimum allowed value for `min_doc_count` should be 1 instead of 0. Having `min_doc_count = 0` might suggest that we return empty buckets too in the response, which is not the case here.

Can we avoid supporting min_doc_count=0 at all? Make 1 the lowest? It's kind of a painful thing in `terms` agg. Both in terms of code and performance.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:12]: Minor grammar nit.
```suggestion
For example, consider the following index:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:47]: Minor grammar nit.
```suggestion
The following aggregation groups documents into buckets. Each bucket identifies a different sub-network. The sub-network is calculated by applying a netmask with prefix length of `24` to each IP address in the `ipv4` field:
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:124]: I'd move the sentence from below up to this section.
```suggestion
If `is_ipv6` is `true`, the response doesn't include a `netmask` for each bucket.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:159]: I'd relocate this above.
```suggestion
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:388]: Can we relocate these sections right above the `Keyed response` heading?

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:166]: This fixes the duplicate colons.
```suggestion
Set the `keyed` flag of `true` to associate an unique IP address key with each bucket and return sub-networks as a hash rather than an array.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:226]: Fixes duplicate colons.
```suggestion
Set the `append_prefix_len` flag to `true` to catenate IP address keys with the prefix length of the sub-network.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:344]: We typically lowercase the literal `ip`. I'd also include an xref.
```suggestion
Document IP address field to aggregate on. The field mapping type must be <<ip,`ip`>>.
```

File: docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc
Comment (jrodewig) [inline docs/reference/aggregations/bucket/ipprefix-aggregation.asciidoc:348]: Minor grammar nits.
```suggestion
Length of the network prefix. For IPv4 addresses, the accepted range is `[0, 32]`. For IPv6 addresses, the accepted range is `[0, 128]`.
```

**Review by jrodewig — APPROVED** [review]
Docs LGTM.

I would suggest moving the `Parameters` and `Response body` sections directly above the `Keyed response` heading. Everything else is non-blocking nits.

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java:156]: In the same line as above, we don't need to allocate a new `BytesRef` or its underlying `byte[]` here.  We can create one (or two, since we might need one for holding `previousSubnet`) and reuse them.  That's a pretty common pattern.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorFactory.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorFactory.java:55]: Passing a null into the Aggregator is the old pattern here.  New aggs should have a no-op implementation instead.  See for example `RangeAggregator.Unmapped`.  Then `IpPrefixAggregator` can assert that its `ValuesSource` is non-null and non-empty.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java
Comment (not-napoleon) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java:1073]: Why are we wrapping `testCase` here?  Does this gain anything?

**Review by not-napoleon — CHANGES_REQUESTED** [review]
I'd like to address the changes around `createUnmapped` and `collect` before merging.  Other comments are not blocking.  Thanks for your work!

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregationBuilder.java
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregationBuilder.java:49]: Yes you are right. Will fix this.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregationBuilder.java
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregationBuilder.java:91]: Right...I will actually also use `<FIELD>.getPreferredName()` into the error message instead of hardcoding the field name.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java
Comment (salvatore-campagna) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java:1073]: This is probably just a result of a refactoring I did and didn't notice wrapping was useless.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java:202]: I think it is not necessary to check it here...

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by salvatore-campagna** [comment]
@elasticmachine test this

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Comment by salvatore-campagna** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregator.java:170]: I still don't think we should be allocating short lived objects here.  I think you can just pass in the `BytesRef` and mutate it in place.

**Review by not-napoleon — COMMENTED** [review]
Two things that I'd still like to fix before merging.  Looking good overall!

**Comment by salvatore-campagna** [comment]
@elasticmachine update branch

**Review by not-napoleon — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
